### PR TITLE
First draft, don't-cover-everything-but-yet-a-good-start-CONTRIBUTING.md

### LINKED ISSUE #468: add Contributors.md
@ariard likely has ideas of what he'd like to see in this

### PR DESCRIPTION
**Pull request by ariard**
A CONTRIBUTING.md, a code formatter, mutation testing. We just have to run Travis on SPARC-32 and we have a real open source project.

Close #468

### CODE REVIEW

**Comment by moneyball** [comment]
@ariard thanks for taking a first pass at this. I did a quick initial review:
- I like the overall tone and sentiment
- We're setting up a Slack for discussion of LDK and Rust-Lightning so once that is completed and public we'll want to mention it here
- There are some grammatical fixes but can wait until we have more feedback from others on content.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+please inform other Lightning implementations diligently.
+
+XXX: (what process ?)
+
+Testing
+-------
+
+Deeply tied with the security aspect, Rust-Lightning developers take testing
+really seriouslt. Due to the modular nature of the project writing new functional
+tests is easy and well-coverage of the codebase is a long-term goal.
```
Comment (jkczyz) [inline CONTRIBUTING.md]: I'd also like to see more unit testing in order to promote less complex, modular design.

Comment (ariard) [reply CONTRIBUTING.md]: Yes add a line on it

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+To contribute a patch, the worflow is a as follows:
+
+  1. Fork Repository
+  1. Create topic branch
+  1. Commit patches
+
+
+In general commits should be atomic and diffs should be easy to read.
+For this reason do not mix ant formatting fixes or code moves with
+actual code changes.
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: Further, each commit, individually, should compile and pass tests, in order to ensure git bisect and other automated tools function properly.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+Anyone may participate in peer review which is expressed by comments in the pull
+request. Typically reviewers will review the code for obvious errors, as well as
+test out the patch set and opine on the technical merits of the patch. PR should
+be reviewed first on the conceptual level before focusing on code style or grammar
+fixes.
+
+Architecture
+------------
+
+XXX: (here or in readme ?)
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: Probably readme, at least in general, though specific "fuzzers live in fuzz/src, with message fuzers in ..." kind of stuff could go here.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+
+Security
+--------
+
+Security is the primary focus of Rust-Lightning, disclosure of security vulnerabilites
+helps prevent user loss of funds. If you think vulnerability is with regard to the spec
+please inform other Lightning implementations diligently.
+
+XXX: (what process ?)
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: Pre-0.1 we can probably say "open github issue". Post-0.1 we should have some list of PGP keys and suggest emailing (to something not on gmail).

**Comment by TheBlueMatt** [comment]
Cool! I left a few high-level comments. Indeed, we'll need someone to go over it for grammar, but I'm probably not the one to do that either lol.

**Comment by TheBlueMatt** [comment]
Also, we should add a line explaining that we use Travis to enforce MSRV of Rust 1.22.0 for the lightning/ workspace (though not for lightning-net-tokio or fuzzers).

**Comment by moneyball** [comment]
Do we want to state "we dont make syscalls" somewhere in the README or CONTRIBUTING doc or elsewhere, so that new contributors understand this important principle of the design?

**Comment by maxgiraldo** [comment]
> Do we want to state "we dont make syscalls" somewhere in the README or CONTRIBUTING doc or elsewhere, so that new contributors understand this important principle of the design?

I think this should live in the README under something like `## Design Goals`. Or if you want to keep the README high level, then say something like "We make no assumptions about the underlying platform"

**Comment by jkczyz** [comment]
> I would also suggest adopting some sort of commit-prefixing protocol, e.g. [conventional commits](https://www.conventionalcommits.org/en/v1.0.0-beta.2/) or similar variant (like what bitcoin core uses). This can help delineate the type of change being made.

Speaking of commits, I'd like to see some guidelines around both commit messages and PR descriptions. When exploring the git log recently, I noticed the *why* for many commits was not given, which makes it difficult to understand the rationale for a change. Sometimes I was able to ascertain this information from the PR description, but often not.

Having clear guidelines about what goes into a commit message and PR description would benefit posterity, including our future selves.

**Comment by maxgiraldo** [comment]
> Having clear guidelines about what goes into a commit message and PR description would benefit posterity, including our future selves.

Linus has a solid [guideline](https://github.com/torvalds/subsurface-for-dirk/blob/master/README.md#contributing).

**Comment by ariard** [comment]
Thanks all for your review, updated with your comments.

@moneyball sorry miss #479, you mention LDK but don't link to any other thinh so take opportunity of this PR to add a link to LDK slack in README.md

File: lightning/src/ln/channelmanager.rs [hunk 5]
```diff
@@ -591,8 +591,7 @@ macro_rules! handle_monitor_err {
 macro_rules! return_monitor_err {
 	($self: ident, $err: expr, $channel_state: expr, $entry: expr, $action_type: path, $resend_raa: expr, $resend_commitment: expr) => {
 		return handle_monitor_err!($self, $err, $channel_state, $entry, $action_type, $resend_raa, $resend_commitment);
-	};
-	($self: ident, $err: expr, $channel_state: expr, $entry: expr, $action_type: path, $resend_raa: expr, $resend_commitment: expr, $failed_forwards: expr, $failed_fails: expr) => {
+	}; ($self: ident, $err: expr, $channel_state: expr, $entry: expr, $action_type: path, $resend_raa: expr, $resend_commitment: expr, $failed_forwards: expr, $failed_fails: expr) => {
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs]: How'd this slip in here?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
+Contributing to Rust-Lightning
+==============================
+
+The Rust-Lightning project operates an open contributor model where anyone is
+welcome to contribute towards development in the form of peer review, documentation,
+testing and patches.
+
+Anyone is invited to contribute without regard to technical experience, "expertise", OSS
+experience age or other social discriminant. Though developing cryptocurrencies demand a
+high-level bar of rigor, adversial thinking, thorough testing and risk-minimization.
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/bar//

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
+Contributing to Rust-Lightning
+==============================
+
+The Rust-Lightning project operates an open contributor model where anyone is
+welcome to contribute towards development in the form of peer review, documentation,
+testing and patches.
+
+Anyone is invited to contribute without regard to technical experience, "expertise", OSS
+experience age or other social discriminant. Though developing cryptocurrencies demand a
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/experience age or social discriminant/experience, age, or other concern/

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+high-level bar of rigor, adversial thinking, thorough testing and risk-minimization.
+Any bug may cost real-money and so impact severely people lifes. That's said we're deeply
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/Any bug may cost real-money and so impact severely people lifes./Any bug may cost users real money./
s/That's said/That said/

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+welcoming people contributing for the first time on an open source project or pick up
+Rust on the ground. Don't be shy, you'll learn.
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/pick up\nRust on the ground/learn\nRust while contributing/g.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+The Rust-Lightning project operates an open contributor model where anyone is
+welcome to contribute towards development in the form of peer review, documentation,
+testing and patches.
+
+Anyone is invited to contribute without regard to technical experience, "expertise", OSS
+experience age or other social discriminant. Though developing cryptocurrencies demand a
+high-level bar of rigor, adversial thinking, thorough testing and risk-minimization.
+Any bug may cost real-money and so impact severely people lifes. That's said we're deeply
+welcoming people contributing for the first time on an open source project or pick up
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/on an open source/to an open source/

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+--------------------
+
+The codebase is maintained using the "contributor workflow" where everyone
+without exception contributes patch proposals using "pull requests". This
+facilitates social contribution, easy testing and peer review.
+
+To contribute a patch, the worflow is a as follows:
+
+  1. Fork Repository
+  1. Create topic branch
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: Why 1. 1. 1.?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+  1. Commit patches
+
+
+In general commits should be atomic and diffs should be easy to read.
+For this reason do not mix ant formatting fixes or code moves with
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/ant/any/

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+actual code changes. Further
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: Stray \n\n here.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+must be given to the long term technical debt. Every new features should
+be covered by functional tests.
+
+When refactoring, structure your PR to make it easy to review and don't
+hesitant to split in multiple small, focused PRs.
+
+The Minimal Supported Rust Version is 1.22.0 (enforced by our Travis).
+
+Commit should expose both issues fixed and solutions rational.
+these [guidelines](https://github.com/torvalds/subsurface-for-dirk/blob/master/README.md#contributing) should be in mind.
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/should be in mind/should be kept in mind.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+------------
+
+Use tabs. If you want to align lines, use spaces. Any desired alignment should
+display fine at any tab-length display setting.
+
+Security
+--------
+
+Security is the primary focus of Rust-Lightning, disclosure of security vulnerabilites
+helps prevent user loss of funds. If you think vulnerability is with regard to the spec
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/If you think vulnerability is with regard to the spec/If you believe the vulnerability may effect other Lightning implementations/

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+please inform other Lightning implementations diligently.
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/diligently//

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+
+Pre-production release, our current process is open an issue on the repo.
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: Maybe with full sentences: Note that rust-lightning is currently considered "pre-production", during this time, there is no special handling of security issues. Please simply open an issue in GitHub.

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+Testing
+-------
+
+Deeply tied with the security aspect, Rust-Lightning developers take testing
+really seriouslt. Due to the modular nature of the project writing new functional
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/really seriouslt/very seriously/

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+tests is easy and well-coverage of the codebase is a long-term goal. Refactoring
```
Comment (TheBlueMatt) [inline CONTRIBUTING.md]: s/well-coverage/good test coverage/
s/long-term/important/

**Comment by ariard** [comment]
Thanks for review, addressed your comments Matt!

**Comment by jkczyz** [comment]
> > Having clear guidelines about what goes into a commit message and PR description would benefit posterity, including our future selves.
> 
> Linus has a solid [guideline](https://github.com/torvalds/subsurface-for-dirk/blob/master/README.md#contributing).

Thanks for passing this along! Some of that is a bit project specific, though. Could we possibly use the same guidelines as Bitcoin Core:

https://chris.beams.io/posts/git-commit/

I'd be up for considering conventional commits (your other suggestion), also. I haven't put much thought into it yet. Do others have experience with them?

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+Peer review
+-----------
+
+Anyone may participate in peer review which is expressed by comments in the pull
+request. Typically reviewers will review the code for obvious errors, as well as
+test out the patch set and opine on the technical merits of the patch. PR should
+be reviewed first on the conceptual level before focusing on code style or grammar
+fixes.
+
+Coding Style
```
Comment (jkczyz) [inline CONTRIBUTING.md]: Could you rename this "Coding Conventions"? Other conventions could be added here or linked to a separate doc in the future.

**Comment by maxgiraldo** [comment]
> I'd be up for considering conventional commits (your other suggestion), also. I haven't put much thought into it yet. Do others have experience with them?

We're piloting a variant of conventional commits at my work. Got some pushback from other developers because it requires some additional work on their part and it's hard to prove the value it provides. Because of this, I think using the same or similar guidelines as Bitcoin Core would be the most beneficial.

File: lightning/src/ln/channelmanager.rs (hunk 5 above)
Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs]: Looks like this is still here.

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+these [guidelines](https://github.com/torvalds/subsurface-for-dirk/blob/master/README.md#contributing) should be kept in mind.
+
+Peer review
+-----------
+
+Anyone may participate in peer review which is expressed by comments in the pull
+request. Typically reviewers will review the code for obvious errors, as well as
+test out the patch set and opine on the technical merits of the patch. PR should
+be reviewed first on the conceptual level before focusing on code style or grammar
+fixes.
```
Comment (jkczyz) [inline CONTRIBUTING.md]: Would it be worth stating the criteria used for merging? Or leave to a follow up?

Comment (ariard) [reply CONTRIBUTING.md]: Good question, @TheBlueMatt should we enforce the at-least 2-ACK for any substantial changes ?

Comment (TheBlueMatt) [reply CONTRIBUTING.md]: I was kinda figuring leave it as a followup - we just dumped 4 new full-time contributors on the project, so I was kinda waiting until folks were up to speed and felt more comfortable with the codebase then we'd revisit it and do something like 2 acks from regular contributors or so (like rust-bitcoin).

**Comment by ariard** [comment]
Updated 0dedb38 with last comments fixed, specially referring to Bitcoin Core git commit convention instead of Linus one.

File: CONTRIBUTING.md (hunk 2 above)
Comment (TheBlueMatt) [reply CONTRIBUTING.md]: Can you add this line?

File: lightning/src/ln/channelmanager.rs (hunk 5 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: Its still here?

Comment (ariard) [reply lightning/src/ln/channelmanager.rs]: Ooops sorry for that, should be removed on last tip

File: CONTRIBUTING.md (hunk 2 above)
Comment (ariard) [reply CONTRIBUTING.md]: Done,

**Comment by ariard** [comment]
Two comments fixed at 34871ea

**Pull request merged**